tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,310 @@
+    }
+
+    // SAFETY: safe to call only once self.initialized() is true
+    unsafe fn get_unchecked(&self) -> &T {
+        &*self.value.with(|ptr| (*ptr).as_ptr())
+    }
+
+    // SAFETY: safe to call only once a permit on the semaphore has been
+    // acquired
+    unsafe fn set_value(&self, value: T) {
+        self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+        self.set_to_initialized();
+        self.semaphore.close();
+    }
+
+    /// Tries to get a reference to the value of the OnceCell.
+    ///
+    /// Returns [`NotInitializedError`] if the value of the OnceCell
+    /// hasn't previously been initialized
+    ///
+    /// [`NotInitializedError`]: crate::sync::NotInitializedError
+    pub fn get(&self) -> Result<&T, NotInitializedError> {
+        if self.initialized() {
+            Ok(unsafe { self.get_unchecked() })
+        } else {
+            Err(NotInitializedError)
+        }
+    }
+
+    /// Sets the value of the OnceCell to the argument value.
+    ///
+    /// If the value of the OnceCell was already set prior to this call, then
+    /// [`AlreadyInitializedError`] is returned
+    ///
+    /// [`AlreadyInitializedError`]: crate::sync::AlreadyInitializedError
+    pub fn set(&self, value: T) -> Result<(), AlreadyInitializedError> {
+        if !self.initialized() {
+            // After acquire().await we have either acquired a permit while self.value
+            // is still uninitialized, or another thread has intialized the value and
+            // closed the semaphore, in which case self.initialized is true and we
+            // don't set the value
+            match self.semaphore.try_acquire() {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        return Ok(());
+                    } else {
+                        unreachable!(
+                            "acquired the permit after OnceCell value was already initialized"
+                        );
+                    }
+                }
+                _ => {
+                    if !self.initialized() {
+                        panic!(
+                            "couldn't acquire a permit even though OnceCell value is uninitialized"
+                        );
+                    }
+                }
+            }
+        }
+
+        Err(AlreadyInitializedError)
+    }
+
+    /// Tries to initialize the value of the OnceCell using the async function `f`.
+    /// If the value of the OnceCell had already been initialized prior to this call
+    /// a reference to that initialized value is returned
+    ///
+    /// This will deadlock if `f` tries to initialize the cell itself.
+    pub async fn get_or_init<F, Fut>(&self, f: F) -> &T
```

Comment:
This documentation is a bit unclear. If the `OnceCell` has not already been initialized, what does it return? (I know the answer, but the doc is unclear.)